### PR TITLE
Always sort logits before nucleus sampling

### PR DESCRIPTION
**Pull request by ivanstepanovftw**
Logits are not sorted before nucleus sampling if TopK is 0 or out of bounds. Reported and solved by @Piezoid https://github.com/ggerganov/llama.cpp/pull/779#issuecomment-1498814682.

Other changes:
Since logits are sorted, first index has maximum probability.

Also remove normalization, because [std::discrete_distribution](https://en.cppreference.com/w/cpp/numeric/random/discrete_distribution) already divides by the sum.

### CODE REVIEW

**Comment by Piezoid** [comment]
The normalization may not be needed, [`std::discrete_distribution`](https://en.cppreference.com/w/cpp/numeric/random/discrete_distribution) already divides by the sum.

**Comment by ivanstepanovftw** [comment]
I can confirm that.

Tested with
```
#include <iostream>
#include <iomanip>
#include <map>
#include <random>
 
int main()
{
    std::random_device rd;
    std::mt19937 gen(rd());
    // std::discrete_distribution<> d({10, 10, 10, 100});
    std::discrete_distribution<> d({100, 100, 100, 1000});
    std::map<int, int> map;
 
    for (int n = 0; n < 1e4; ++n)
        ++map[d(gen)];
 
    for(const auto& [num, count] : map)
        std::cout << num << " generated " << std::setw(4) << count << " times\n";
}
```

**Comment by Piezoid** [comment]
On a related note, I tried an optimization (which is getting out of scope for this PR): during the initial pass over logits and the application of temperature scaling, there's a way to bypass tokens that have probabilities much lower than the highest probability using a heuristic:
```c++
const float maxl = *std::max_element(plogits, plogits + n_logits);
const float minp_ratio = 1e-6; // Ratio lowest admissible prob : highest prob
const float minl = maxl / repeat_penalty + log(minp_ratio) * temp; // Assumes the best token is penalized
// Then, when plogits[i] < minl, don't push_back the pair(logit, i) into logits_id
```
Sampling is clearly not the most compute heavy task, so it might not be a good idea to introduce a heuristic that could potentially fail. I was experimenting with a more complex repetition search, that could justify it.

**Review by ggerganov — APPROVED** [review]
